[Fixes #2459] Do not use a variable substitution when updating python.pythonPath.

ericsnowcurrently opened the pull request:
(for #2459)

- [x] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR)
- [x] Title summarizes what is changing
- [x] Has a [news entry](https://github.com/Microsoft/vscode-python/tree/master/news) file (remember to thank yourself!)
- [x] Has unit tests & system/integration tests
- [x] ~~Any new/changed dependencies in [`package.json`](https://github.com/Microsoft/vscode-python/blob/master/package.json) are pinned (e.g. `"1.2.3"`, not `"^1.2.3"` for the specified version)~~
- [x] ~~[`package-lock.json`](https://github.com/Microsoft/vscode-python/blob/master/package-lock.json) has been regenerated by running `npm install` (if dependencies have changed)~~

DonJayamanne commented on src/client/interpreter/configuration/services/workspaceFolderUpdaterService.ts:
```diff
@@ -1,22 +1,26 @@
-import * as path from 'path';
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Please remove the copyright, this is to be added only for new files.

ericsnowcurrently replied:
fixed

DonJayamanne commented on src/client/interpreter/configuration/services/workspaceFolderUpdaterService.ts:
```diff
@@ -1,22 +1,26 @@
         if (pythonPathValue && pythonPathValue.workspaceFolderValue === pythonPath) {
             return;
         }
-        if (pythonPath.startsWith(this.workspaceFolder.fsPath)) {
```
We need to restore the old change, and add the relative path. Right now it will add the fully qualified path, this makes it less portable for teams.

DonJayamanne replied:
The benefit of using `${workspaceFolder}` in settings was teams could share the `settings.json` file as the paths are relative, now they are absolute making them less portable, this would be a breaking change.

ericsnowcurrently replied:
The whole point of this PR is to remove the variable substitution.  I'm not clear what alternative you are suggesting.

ericsnowcurrently replied:
fixed

DonJayamanne commented on src/client/interpreter/configuration/services/workspaceUpdaterService.ts:
```diff
@@ -1,22 +1,26 @@
         if (pythonPathValue && pythonPathValue.workspaceValue === pythonPath) {
             return;
         }
-        if (pythonPath.startsWith(this.workspace.fsPath)) {
```
Please see previous comment.

DonJayamanne commented on src/client/interpreter/configuration/services/workspaceUpdaterService.ts:
```diff
@@ -1,22 +1,26 @@
-import * as path from 'path';
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
```
Same as previous comment.

DonJayamanne commented on src/test/interpreter/configuration/services/workspaceUpdaterService.unit.test.ts:
```diff
@@ -0,0 +1,131 @@
+    if (filename === '') {
+        return '';
+    }
+    const parts = filename.split('/');
```
Do not hardccode file path seprators. On Windows its `\` and others its `/`

ericsnowcurrently replied:
In tests it's much more readable to do it the way I've done it with little cost.  The point of this helper function is to keep the tests cross-platform while keeping the test code more readable.

DonJayamanne replied:
The fact that you are splitting with '/' doesn't make it cross platform.
On windows path separators are `\`.

ericsnowcurrently replied:
The point is that the tests define everything as if on linux and then `normalizeFilename()` fixes up the filenames to be cross-platform.  That makes the tests more readable and easier to write.

DonJayamanne commented on src/test/interpreter/configuration/services/workspaceUpdaterService.unit.test.ts:
```diff
@@ -0,0 +1,131 @@
+    WorkspacePythonPathUpdaterService
+} from '../../../../client/interpreter/configuration/services/workspaceUpdaterService';
+
+function normalizeFilename(filename: string): string {
```
Please consider using `path.normalize`.

ericsnowcurrently replied:
This isn't about normalizing the path in the normal sense.  So `path.normalize()` isn't a good fit.  See my other comment.

DonJayamanne replied:
Then what are we doing? I just don't see the purpose of this, when we can hard code the expected value in the array, doing that makes it completely unnecessary to write more code in the tests, and makes it a whole lot easier to read and understand.

DonJayamanne commented on src/test/interpreter/configuration/services/workspaceUpdaterService.unit.test.ts:
```diff
@@ -0,0 +1,131 @@
+
+    function setExpected(expected: string) {
+        config.setup(c => c.update('pythonPath', expected, false))
+            .returns(() => Promise.resolve());
```
You need to add a `.verifiable(TypeMoq.Times.xxx)` in order to use this for verification.
Else `verifyAll` does nothing.

ericsnowcurrently replied:
Note that I've create the mocks in "strict" mode.  This means that they are added for verification automatically.

DonJayamanne commented on src/test/interpreter/configuration/services/workspaceUpdaterService.unit.test.ts:
```diff
@@ -0,0 +1,131 @@
+        ['on $PATH', 'project', 'python3', ''],
+        ['under $HOME', 'project', '~/my-venv/bin/python3', ''],
+
+        ['relative -- at workspace root', 'project', 'project/python3', ''],
```
You might want to use `path.join` instead of hardcoding `/`, else you're testing only against unix paths.

ericsnowcurrently replied:
The tests are cross-platform.  See the comments above relative to the helper.

DonJayamanne commented on src/test/interpreter/configuration/services/workspaceUpdaterService.unit.test.ts:
```diff
@@ -0,0 +1,131 @@
+        ['absolute - starts with workspace root', '/home/user/project', '/home/user/project/my-venv/bin/python3', ''],
+        ['absolute - does not start with workspace root', '/home/user/project', '/home/user/my-venv/bin/python3', '']
+    ];
+    for (let [testName, workspaceRoot, pythonPath, expected] of tests) {
```
In all of the items in the array, the 4th item is always `''`, meaning, `expected` is always empty, hence I do not see the purpose of the `normalizeFilename`.
Also why can't we write expected file name in the test rather than computing it, more specific.

ericsnowcurrently replied:
I made `''` mean "use `pythonPath` solely so the array is less cluttered and to make it extra clear that pythonPath does not get changed.

DonJayamanne replied:
Don't understand you, either way, why can't we remove it!
It is ALWAYS empty, hence see no point in having more code to check if its empty or not, when it is.

ericsnowcurrently replied:
It's not always empty though. :)

DonJayamanne commented on src/test/interpreter/configuration/services/workspaceUpdaterService.unit.test.ts:
```diff
@@ -0,0 +1,131 @@
+        if (expected === '') {
+            expected = pythonPath;
+        } else {
+            expected = normalizeFilename(expected);
```
This code will never get executed.

ericsnowcurrently replied:
Sure, with the current test cases it won't.  However, I anticipated that there would be other test cases that would change the path, hence accommodating separate "expected" paths in the tests.

DonJayamanne replied:
I'd say this is YAGNI, lets not plan for the future, its easy to add it then, hence lets remove dead code now.

ericsnowcurrently replied:
Except I just had to add tests that used it. :)

DonJayamanne replied:
Personally I don't see the point of having `normalizeFilename`.
We can easily hard code the strings as you have already done in the array. 

I'm looking at this code, and I cannot understand why we cannot just hardcode the expected value as the 4th item in the index, that increases readability.
Right now, one has to follow the code, look at why its empty, then check the if condition, then follow and understand what normalizeFileName does, all to understand what the expected name is!!

DonJayamanne replied:
Please remove tests for CWD, its impossible scenario.
I.e. we cannot predict outcome for an impossible scenario.
I didn't bring this up earlier as i assumed this was the workspace folder, now understand this based on our conversation from earlier today.

Now, looking at the revised code, I personally don't see the need for new unit test file either, we should be able to write 2 new tests in the old test file that will account for **non-relative paths**.

I didn't raise this as the code had to be refactored to bring back the old code, now that the old functionality has been brought back to store relative paths, this makes testing easier.

Also the other tests are much simpler than this (this feels more complicated with `normalize...` and its hardcoded for `unix` paths).

Summary:
* Add tests for pythonpath = 'python3'
* Add tests for pythonpath = 'home/usr/dev/env/bin/python'
(make use of path.join(...) instead of hard coding `/`.
This should simplify the tests.

You should be able to use the same concept of using a loop with an array of items containing the data and the expectations.

DonJayamanne commented on src/test/interpreter/configuration/services/workspaceFolderUpdaterService.unit.test.ts:
```diff
@@ -0,0 +1,131 @@
+    WorkspaceFolderPythonPathUpdaterService
+} from '../../../../client/interpreter/configuration/services/workspaceFolderUpdaterService';
+
+function normalizeFilename(filename: string): string {
```
Not sure what this is doing, either way we have code duplicated.
Please try to use `path.normalize`.
Also, please do not hardcode `/`, this helps with cross platform tests, specifically windows.

ericsnowcurrently replied:
See other comments about this.

DonJayamanne commented on src/test/interpreter/configuration/services/workspaceFolderUpdaterService.unit.test.ts:
```diff
@@ -0,0 +1,131 @@
+        if (expected === '') {
+            expected = pythonPath;
+        } else {
+            expected = normalizeFilename(expected);
```
See comment in next file, this line of code will never get executed.

DonJayamanne replied:
Dead code.

DonJayamanne commented on src/test/interpreter/configuration/services/workspaceFolderUpdaterService.unit.test.ts:
```diff
@@ -0,0 +1,131 @@
+
+    const tests: [string, string, string, string][] = [
+        // (test name, root, new pythonPath, expected)
+        ['on $PATH', 'project', 'python3', ''],
```
Just as in other file, why is the 4th argument computed. Why can't we hard code it here e.g. `['on $PATH', 'project', 'python3', 'python3'],`
`['under $HOME', 'project', '~/my-venv/bin/python3', '~/my-venv/bin/python3'],`
`['relative -- under workspace root', 'project', 'project/my-venv/bin/python3', 'my-venv/bin/python3'],`

ericsnowcurrently commented:
I've simplified things by factoring out a base class for the two `*Service` classes.  I've also added more tests and dropped `normalizeFilename()`.  FWIW, I still find the composed (via `path.join()`) filenames less readable, but this isn't important enough to block on. :)

DonJayamanne commented:
Why the whole refactor for this simple change? Initially nothing much changed, now you have refactored pretty much the whole thing!!

DonJayamanne commented on src/client/common/configSettings.ts:
```diff
@@ -28,6 +28,41 @@ const untildify = require('untildify');
 
 export const IS_WINDOWS = /^win/.test(process.platform);
 
+export function getSetting<T>(
```
This function is unnecessary.

DonJayamanne replied:
More code to be tested, this is basically a wrapper for a one-line code in VSC.

DonJayamanne commented:
Please review my PR and the changes https://github.com/Microsoft/vscode-python/pull/2705
Lets discuss tomorrow.

The new function `getSetting ` you introduced basically replaces the following code:
```
const pythonPathValue = pythonConfig.inspect<string>('pythonPath');	
```

The new function `matchSetting ` you introduced basically replaces the following code:
```
const pythonPathValue = pythonConfig.inspect<string>('pythonPath');	
if (pythonPathValue && pythonPathValue.workspaceValue === pythonPath) {
    return;
}	     
```

ericsnowcurrently commented:
I'm fine with just landing the minimal change.  The point of all the extra code is to simplify things and add unit test coverage, both of which are worth doing.  I considered putting that extra stuff in a separate PR, but figured it wasn't that big a deal.  So in the interest of getting the actual fix landed I'll punt on that.

FWIW, I'd like to discuss your objections separately (not here or now) as I have a different perspective on the value of the extra changes.
